File: ios/Share/ShareVC.m
//
// ShareViewController.m
// Share
//
// Created by Marcus Rohrmoser on 18.03.15.
// Copyright (c) 2015 Marcus Rohrmoser. All rights reserved.
//

#import "ShareVC.h"
#import <MobileCoreServices/UTCoreTypes.h>
#import "NSUserDefaults+Share.h"
#import "ShaarliM.h"

@interface ShareVC() <UITextFieldDelegate, UITextViewDelegate, NSURLSessionDelegate> {
    SLComposeSheetConfigurationItem *itemTitle;
    SLComposeSheetConfigurationItem *itemTags;
    SLComposeSheetConfigurationItem *itemPrivate;
}
@property (readonly, strong, nonatomic) ShaarliM *shaarli;

@end

@implementation ShareVC

-(void)viewDidLoad
{
    MRLogD(@"-", nil);
    [super viewDidLoad];
    if( !self.shaarli ) {
        _shaarli = [[ShaarliM alloc] init];
        [self.shaarli load];
        NSParameterAssert(self.shaarli.title);
    }
}


-(NSArray *)configurationItems
{
    MRLogD(@"-", nil);
    // To add configuration options via table cells at the bottom of the sheet, return an array of SLComposeSheetConfigurationItem here.


    itemTitle = [[SLComposeSheetConfigurationItem alloc] init];
    itemTitle.title = NSLocalizedString(@"Title", @"ShaareVC");
    itemTitle.value = self.contentText;
    [itemTitle setTapHandler:^(void) {
         MRLogD (@"", nil);
     }
    ];

    itemTags = [[SLComposeSheetConfigurationItem alloc] init];
    itemTags.title = NSLocalizedString (@"Tags", @"ShaareVC");
    itemTags.value = @"";
    [itemTags setTapHandler:^(void) {
         MRLogD (@"", nil);
     }
    ];

    itemPrivate = [[SLComposeSheetConfigurationItem alloc] init];
    [itemPrivate setTitle:NSLocalizedString (@"Private", @"ShaareVC")];
    itemPrivate.value = NSLocalizedString (@"Private", @"ShaareVC");
    __weak typeof (itemPrivate)wr = itemPrivate;
    [itemPrivate setTapHandler:^(void) {
         MRLogD (@"", nil);
         const BOOL priv = [NSLocalizedString (@"Private", @"ShaareVC") isEqualToString:wr.value];
         wr.value = priv ? NSLocalizedString (@"Public", @"ShaareVC"):NSLocalizedString (@"Private", @"ShaareVC");
     }
    ];

#if 0
    SLComposeSheetConfigurationItem *itemShaar = [[SLComposeSheetConfigurationItem alloc] init];
    itemShaar.title = NSLocalizedString (@"Shaarə", @"ShaareVC");
    itemShaar.value = self.shaarli.title;
    [itemShaar setTapHandler:^(void) {
         MRLogD (@"", nil);
         NSURL *b = [NSURL URLWithString:SELF_URL_PREFIX @"://command/"];
         NSURL *c = [NSURL URLWithString:@"./https://google.com?q=a&a=b#c" relativeToURL:b];
         [self.extensionContext openURL:c completionHandler:^(BOOL success) {
              MRLogD (@"%d %@", success, c.absoluteString, nil);
          }
         ];
     }
    ];
    return @[itemTitle, itemTags, itemPrivate, itemShaar];
#else
    return @[itemTitle, itemTags, itemPrivate];
#endif
}


-(void)viewWillAppear:(BOOL)animated
{
    MRLogD(@"%@", [NSUserDefaults shaarliDefaults], nil);
    [super viewWillAppear:animated];
    self.view.tintColor = [UIColor colorWithRed:128 / 255.0f green:173 / 255.0f blue:72 / 255.0f alpha:1.0f];
    NSParameterAssert(self.shaarli);

    self.title = @"Shaarə"; // self.shaarli.title;
}


-(void)viewDidAppear:(BOOL)animated
{
    MRLogD(@"-", nil);
    [super viewDidAppear:animated];
    // self.payload[@"title"] = [self.extensionContext.inputItems[0] attributedContentText].string;
    NSParameterAssert(itemTitle);
    itemTitle.value = self.contentText;
    // self.textView.text = @"";
}


-(void)presentationAnimationDidFinish
{
    MRLogD(@"we may need to update the display.", nil);
}


-(BOOL)isContentValid
{
    MRLogD(@"-", nil);
    // Do validation of contentText and/or NSExtensionContext attachments here
    return YES;
}


-(void)didSelectPost
{
    MRLogD(@"-", nil);
    // This is called after the user selects Post. Do the upload of contentText and/or NSExtensionContext attachments.

    NSParameterAssert(1 == self.extensionContext.inputItems.count);
    NSExtensionItem *item = self.extensionContext.inputItems[0];
    // MRLogD(@"title %@", item.attributedTitle.string, nil);
    MRLogD(@"content %@", item.attributedContentText.string, nil);
    NSDictionary *ui = item.userInfo;
    __block NSURL *url = nil;
    for( NSItemProvider *itemProvider in item.attachments ) {
        NSString *t = (__bridge NSString *)kUTTypeText;
        MRLogD(@"registeredTypeIdentifiers %@", itemProvider.registeredTypeIdentifiers, nil);
        if( [itemProvider hasItemConformingToTypeIdentifier:t] )
            [itemProvider loadItemForTypeIdentifier:t options:nil completionHandler:^(NSURL * url_, NSError * error) {
                 MRLogD (@"huhu %@", url_, nil);
                 NSAssert (NO, @"Yikes!", nil);
                 url = url_;
             }
            ];
    }

    NSString *confName = [[[[NSBundle bundleForClass:[self class]] infoDictionary] valueForKey:@"CFBundleIdentifier"] stringByAppendingString:@".backgroundpost"];
    NSURLSessionConfiguration *conf = [NSURLSessionConfiguration backgroundSessionConfigurationWithIdentifier:confName];
    NSURLSession *session = [NSURLSession sessionWithConfiguration:conf delegate:self delegateQueue:nil];

    // [self.shaarli postURL:<#(NSURL *)#> title:<#(NSString *)#> tags:<#(id<NSFastEnumeration>)#> description:<#(NSString *)#> private:<#(BOOL)#> session:session completion:<#^(ShaarliM *me, NSError *error)completion#>]

    // Inform the host that we're done, so it un-blocks its UI. Note: Alternatively you could call super's -didSelectPost, which will similarly complete the extension context.
    [self.extensionContext completeRequestReturningItems:@[item] completionHandler:^(BOOL expired) {
         MRLogD (@"-", nil);
     }
    ];
}


-(void)didSelectCancel
{
    MRLogD(@"-", nil);
    NSError *error = [NSError errorWithDomain:NSCocoaErrorDomain code:NSUserCancelledError userInfo:nil];
    [self.extensionContext cancelRequestWithError:error];
}


#pragma mark NSURLSessionDelegate


-(void)URLSession:(NSURLSession *)session didBecomeInvalidWithError:(NSError *)error
{
    MRLogD(@"", nil);
}


@end

File: swift4/Share/ShareVC.swift
//
//  ShareVC.swift
//  Share
//
//  Created by Marcus Rohrmoser on 02.03.20.
//  Copyright © 2020-2020 Marcus Rohrmoser mobile Software http://mro.name/me. All rights reserved.
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

import UIKit
import Social
import MobileCoreServices

fileprivate func stringFromPrivacy(_ priv : Bool) -> String
{
    return priv
        ? NSLocalizedString("Private 🔐", comment:"ShareVC")
        : NSLocalizedString("Public 🔓", comment:"ShareVC")
}

fileprivate func privacyFromString(_ s : String) -> Bool
{
    return s != stringFromPrivacy(false)
}

@objc (ShareVC) // https://blog.hellocode.co/post/share-extension/
class ShareVC: SLComposeServiceViewController {

    var current         : BlogM?
    var wasTouched      = false
    var itemTitle       : SLComposeSheetConfigurationItem?
    var itemAudience    : SLComposeSheetConfigurationItem?
    var ctx             : HtmlFormDict = [:]
    var url             : URL = URLEmpty

    override func viewDidLoad() {
        debugPrint("viewDidLoad")
        super.viewDidLoad()
    }

    override func configurationItems() -> [Any]! {
        debugPrint("configurationItems")
        guard let iTi = SLComposeSheetConfigurationItem() else {return []}
        iTi.title = NSLocalizedString("Title", comment:"ShareVC")
        iTi.value = self.contentText

        guard let iAu = SLComposeSheetConfigurationItem() else {return []}
        iAu.title = NSLocalizedString("Audience", comment:"ShareVC")
        iAu.value = stringFromPrivacy(false)
        weak var wself = self
        iAu.tapHandler = {
            guard let sf = wself else {return}
            guard let iAu = sf.itemAudience else {return}
            iAu.value = stringFromPrivacy( !privacyFromString(iAu.value) )
            sf.wasTouched = true
        }

        itemTitle = iTi
        itemAudience = iAu
        return [iTi, iAu]
    }

    override func viewWillAppear(_ animated: Bool) {
        debugPrint("viewWillAppear")
        super.viewWillAppear(animated)
        view.tintColor = UIColor(red:128 / 255.0, green:173 / 255.0, blue:72 / 255.0, alpha:1.0)
        assert(itemTitle != nil)
        assert(itemAudience != nil)

        let sha = ShaarliM.shared
        current = sha.loadBlog(sha.defaults)
        guard let current = current else {return}
        let c = ShaarliHtmlClient()

        textView.keyboardType = .twitter

        guard let itemTitle = itemTitle else {return}
        guard let itemAudience = itemAudience else {return}
        guard let textView = textView else {return}

        title = current.title
        itemTitle.value = contentText
        let tagsDefault = current.tagsActive
            ? current.tagsDefault
            : ""

        let txt = "" != tagsDefault
            ? "\(tagsDefault) "
            : ""
        textView.text = "\(txt)\(NSLocalizedString("🔄", comment:"ShareVC"))"
        itemAudience.value = stringFromPrivacy(current.privateDefault)

        let tUrl = kUTTypeURL as String
        let tTxt = kUTTypeText as String
        weak var ws = self
        for _item in (extensionContext?.inputItems)! {
            let item = _item as! NSExtensionItem
            for _ip in (item.attachments!) {
                let ip = _ip as! NSItemProvider // required for Xcode <10
                // see predicate from http://stackoverflow.com/a/27932776
                if( ip.hasItemConformingToTypeIdentifier(tUrl) ) {
                    ip.loadItem(forTypeIdentifier:tUrl, options:nil) { (_url, err) in
                        guard let ws = ws else {return}
                        guard let _url = _url as? URL else {
                            return
                        }
                        guard let err = err else {
                            c.get(current.endpoint, _url, { (ctx, _url, tit, dsc, tgs, pri, err) in
                                let r = tagsNormalise(description:tit, extended:dsc, tags:tgs, known:[])
                                DispatchQueue.main.async {
                                    ws.ctx = ctx
                                    ws.url = _url
                                    itemTitle.value = "" != r.description
                                        ? r.description
                                        : itemTitle.value
                                    textView.text = "" != r.extended
                                        ? r.extended
                                        : txt
                                    itemAudience.value = stringFromPrivacy(pri)
                                }
                            })
                            return
                        }
                        ws.showError(
                            title:NSLocalizedString("Error", comment: ""),
                            message:err.localizedDescription
                        )
                    }
                }
                if( ip.hasItemConformingToTypeIdentifier(tTxt) ) {
                    ip.loadItem(forTypeIdentifier:tTxt, options:nil) { (_txt, err) in
                        guard let ws = ws else {return}
                        guard let err = err else {
                            debugPrint("done. title:\(itemTitle.value ?? "-") txt:\(String(describing: _txt))")
                            return
                        }
                        ws.showError(
                            title:NSLocalizedString("Error", comment: ""),
                            message:err.localizedDescription
                        )
                    }
                }
            }
        }
    }

    override func didSelectPost() {
        debugPrint("didSelectPost")
        guard let current = current else {return}
        let c = ShaarliHtmlClient()
        guard let tit = itemTitle?.value else {return}
        guard let dsc = textView.text else {return}
        let pri = privacyFromString((itemAudience?.value)!)
        let r = tagsNormalise(description:tit, extended:dsc, tags:[], known:[])
        c.add(current.endpoint, ctx, url, r.description, r.extended, r.tags, pri) {
            if $0 != "" {
                self.showError(
                    title:NSLocalizedString("Error", comment: "ShareVC"),
                    message:$0
                )
            }
            super.didSelectPost()
        }
        // Inform the host that we're done, so it un-blocks its UI. Note: Alternatively you could call super's -didSelectPost, which will similarly complete the extension context.
        // self.extensionContext!.completeRequest(returningItems: [], completionHandler: nil)
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        if nil != current {
            wasTouched = false
            return
        }

        let alert = UIAlertController(
            title:NSLocalizedString("No Shaarli found", comment:"ShareVC"),
            message: NSLocalizedString("Please add one in the Shaarli💫 settings.", comment:"ShareVC"),
            preferredStyle:.alert
        )

        alert.addAction(UIAlertAction(
            title: NSLocalizedString("Cancel", comment:"ShareVC"),
            style:.cancel,
            handler:{ (_) in
                self.cancel()
            }
        ))
        alert.addAction(UIAlertAction(
            title: NSLocalizedString("Settings", comment:"ShareVC"),
            style:.default,
            handler:{ (_) in
                // https://stackoverflow.com/a/44499222/349514
                DispatchGroup().notify(queue: DispatchQueue.main) {
                    let _ = self.openURL(URL(string:"\(SELF_URL_PREFIX):///settings")!)
                }
                self.cancel()
            }
        ))
        DispatchQueue.main.async {
            self.present(alert, animated:true, completion:nil)
        }
    }

    fileprivate func showError(title:String, message:String) {
        let alert = UIAlertController(
            title:title,
            message:message,
            preferredStyle:.alert
        )
        alert.addAction(UIAlertAction(
            title:title,
            style:.cancel,
            handler:{ (_) in
                self.cancel()
            }
        ))

        DispatchQueue.main.async {
            self.present(alert, animated:true, completion:nil)
        }
    }
    
    override func presentationAnimationDidFinish() {
        debugPrint("presentationAnimationDidFinish")
    }

    override func isContentValid() -> Bool {
        debugPrint("isContentValid")
        // Do validation of contentText and/or NSExtensionContext attachments here
        wasTouched = true
        return true
    }

    // No preview image right upper inside the share dialog.
    override func loadPreviewView() -> UIView! {
        return nil
    }

    override func didSelectCancel() {
        debugPrint("didSelectCancel")
        super.didSelectCancel()
    }

    // https://stackoverflow.com/a/44499222/349514
    // Function must be named exactly like this so a selector can be found by the compiler!
    // Anyway - it's another selector in another instance that would be "performed" instead.
    @objc fileprivate func openURL(_ url: URL) -> Bool {
        var rep: UIResponder? = self
        while rep != nil {
            if let app = rep as? UIApplication {
                return app.perform(#selector(openURL(_:)), with: url) != nil
            }
            rep = rep?.next
        }
        return false
    }
}